### PR TITLE
add Min-K%++ attack

### PR DESCRIPTION
**Pull request by zjysteven**
Hi @iamgroot42, first thanks for putting up this unified benchmark. 

## Overview
This pull request add our proposed method, Min-K%++, which outperforms other reference-free methods and performs on par with the Ref method on MIMIR. See below experiment part for details.

## Changed files
We follow the instruction to create `attacks/mink_plus_plus.py` and register the method in `attacks/all_attacks.py` and `attacks/utils.py`. Our method needs the full probabilities over the vocabulary, so we modify the `get_probabilities` function in `models.py` with an additional argument `return_all_probs` to return full probs.

Meanwhile, we change `run.py` accordingly to include our Min-K%++ attack. Additionally, we let both Min-K% and Min-K%++ to run over a list of `k` values, which allows both methods to be compared by their upper bounds.

Lastly, we change the readme to incorporate Min-K%++.

## Experiments
image
I'm putting this result table here in case MIMIR wants to have a leaderboard or something. We are using non-deduped pythia models and the default `ngram_13_0.8` setting. The data splits are fetched from `cache_100_200_1000_512` folder on huggingface.

Let us know if there are any issues. Otherwise, we look forward to Min-K%++ being integrated into the MIMIR benchmark.

### CODE REVIEW

File: mimir/attacks/min_k_plus_plus.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        and returns normalized likelihood when computed over top k% of ngrams.
+        """
+        # Hyper-params specific to min-k attack
+        k: float = kwargs.get("k", 0.2)
+        all_probs = kwargs.get("all_probs", None)
+
+        target_prob, all_prob = (
+            (probs, all_probs)
+            if (probs is not None and all_probs is not None)
+            else self.model.get_probabilities(document, tokens=tokens, return_all_probs=True)
```
Comment (iamgroot42) [inline mimir/attacks/min_k_plus_plus.py:30]: Is there a reason for computing softmax and log_softmax separately? Given access to the latter (which is default right now), a exp() operation should give you the desired softmax outputs for use in mink++?

File: run.py [hunk 2]
```diff
@@ -150,17 +150,38 @@ def get_mia_scores(
...
-                            probs=s_tk_probs,
-                            detokenized_sample=(
-                                detokenized_sample[i]
-                                if config.pretokenized
-                                else None
-                            ),
-                            loss=loss,
-                        )
-                        sample_information[attack].append(score)
+                        if attack in [AllAttacks.MIN_K, AllAttacks.MIN_K_PLUS_PLUS]:
```
Comment (iamgroot42) [inline run.py:153]: Our intent with the structure was to have as few attack-specific checks as possible. We plan on adding attack-specific hyper-parameters in the config file, so that any changes required for the attack here can be achieved by different config files, and not hard-coded hyper-parameters and switch-case in code!

**Comment by iamgroot42** [comment]
Thanks a lot for the PR, @zjysteven ! I have a couple suggestions that I have added via comments. The fact that you submitted such a well-structured PR is alone sufficient and I would understand if you would rather have us make these changes, but would be great if you could make those changes

**Comment by zjysteven** [comment]
Yes I can make further changes. Somehow I'm not seeing your comments. Would you please point me to the comments?

**Review by iamgroot42 — CHANGES_REQUESTED** [review]

File: mimir/attacks/min_k_plus_plus.py (hunk 1 above)
Comment (zjysteven) [reply mimir/attacks/min_k_plus_plus.py:30]: I see. Earlier I misread and thought that we have softmax, and since log_softmax is numerically more stable than log(softmax) I compute them separately (but yeah as you pointed what we have already is log_softmax). I will change this so the forward pass of the model (`get_probabilites`) will be kept as-is.

File: run.py (hunk 2 above)
Comment (zjysteven) [reply run.py:153]: Also a valid point. I will just remove this modification so `run.py` follows the current design.

File: mimir/attacks/min_k_plus_plus.py (hunk 1 above)
Comment (zjysteven) [reply mimir/attacks/min_k_plus_plus.py:31]: I do want to clarify here that while `target_prob` is the log probability of each input token, `all_prob` is the log probabilities of every token in the vocabulary (the whole categorical distribution).

**Comment by zjysteven** [comment]
I've made changes accordingly in the latest commit. Would you take another look?

File: mimir/models.py [hunk 3]
```diff
@@ -116,7 +113,6 @@ def get_probabilities(self,
                 if no_grads:
                     logits = logits.cpu()
                 shift_logits = logits[..., :-1, :].contiguous()
-                probabilities = torch.nn.functional.softmax(shift_logits, dim=-1)
```
Comment (zjysteven) [inline mimir/models.py]: Like suggested here we don't need to separately call `softmax` given that `log_softmax` will be computed anyway.

**Comment by iamgroot42** [comment]
Thanks for making the changes! Will run our standard workflow and merge it

**Review by iamgroot42 — APPROVED** [review]
